internal/lsp/lsproto/jsonrpc.go

```diff
@@ -43,56 +43,88 @@ func (id *ID) UnmarshalJSON(data []byte) error {
 	return json.Unmarshal(data, &id.int)
 }
 
-// TODO(jakebailey): NotificationMessage? Use RequestMessage without ID?
+type Message struct {
+	JSONRPC JSONRPCVersion `json:"jsonrpc"`
+}
+
+type NotificationMessage struct {
+	Message
+	Method Method `json:"method"`
+	Params any    `json:"params"`
+}
 
 type RequestMessage struct {
-	JSONRPC JSONRPCVersion `json:"jsonrpc"`
-	ID      *ID            `json:"id"`
-	Method  Method         `json:"method"`
-	Params  any            `json:"params"`
+	Message
+	ID     *ID    `json:"id"`
+	Method Method `json:"method"`
+	Params any    `json:"params"`
 }
 
-func (r *RequestMessage) UnmarshalJSON(data []byte) error {
+type RequestOrNotificationMessage struct {
+	NotificationMessage *NotificationMessage
+	RequestMessage      *RequestMessage
+}
+
+func (r *RequestOrNotificationMessage) UnmarshalJSON(data []byte) error {
 	var raw struct {
 		JSONRPC JSONRPCVersion  `json:"jsonrpc"`
-		ID      *ID             `json:"id"`
+		ID      *ID             `json:"id,omitempty"`
```

Comment:
Generally speaking, we should be using `omitzero` (new in Go 1.24) rather than `omitempty`.

Reply:
There are other of these, FWIW, I just commented on the one.

Reply:
I went ahead and normalized this file to always use omitzero!